migen/build/platforms/tinyfpga_a.py

```diff
@@ -0,0 +1,155 @@
+    # if it doesn't match one of the below frequencies. Since we tend to deal
+    # with periods in nanoseconds, choose the closest frequency (in MHz)
+    # within +/-5%. These numbers were extracted from
+    # "MachXO2 sysCLOCK PLL Design and Usage Guide"
+    supported_freqs = [
+        2.08, 2.15, 2.22, 2.29, 2.38, 2.46, 2.56, 2.66, 2.77, 2.89,
+        3.02, 3.17, 3.33, 3.50, 3.69, 3.91, 4.16, 4.29, 4.43, 4.59,
+        4.75, 4.93, 5.12, 5.32, 5.54, 5.78, 6.05, 6.33, 6.65, 7.00,
+        7.39, 7.82, 8.31, 8.58, 8.87, 9.17, 9.50, 9.85, 10.23, 10.64,
+        11.08, 11.57, 12.09, 12.67, 13.30, 14.00, 14.78, 15.65, 15.65, 16.63,
+        17.73, 19.00, 20.46, 22.17, 24.18, 26.60, 29.56, 33.25, 38.00, 44.33,
+        53.20, 66.50, 88.67, 133.00
+    ]
+
+    def __init__(self, period, out):
+        freq_str = "{:.2f}".format(self.nearest_freq(period))
+        self.specials += Instance("OSCH",
+                                  p_NOM_FREQ=freq_str,
+                                  i_STDBY=C(0),
+                                  o_OSC=out
+                                  )
+
+    def nearest_freq(self, period):
+        target_freq = 1000.0/period
+
+        # A None match is defined to not be in range. Handles special cases
+        # described below.
+        def in_range(actual, err=0.05):
+            return abs((target_freq - actual)/target_freq) <= err if actual else False
+
+        # Agreed-upon algorithm:
+        # https://github.com/m-labs/migen/pull/111#discussion_r189687411
+        i = bisect.bisect_right(self.supported_freqs, target_freq)
+
+        # Special case: desired freq was < 2.08, but could still within +5%
+        lo_match = self.supported_freqs[i-1] if i else None
+
+        # Special case: desired freq was > 133.00, but could still within -5%
+        hi_match = self.supported_freqs[i] if i < len(self.supported_freqs) else None
+
+        if in_range(lo_match):
+            selected_freq = lo_match
+        elif in_range(hi_match):
+            selected_freq = hi_match
+        else:
+            raise ValueError("Clock period out of range (-/+5%) for internal oscillator.")
+
+        return selected_freq
+
+
+class OschRouting(Module):
+    def __init__(self):
+        self.osch_used = False  # Only one default clock,
+        # even if > 1 OSCH is possible, so run only once.
+        self.mach_clk_sig = Signal()
+
+    def mk_clk(self, name, clk_period):
+        if not self.osch_used:
+            self.mach_clk_sig.name_override = name
+            self.submodules.mclk = MachClock(clk_period, self.mach_clk_sig)
+            self.osch_used = True
+        else:
+            raise ConstraintError
+        return self.mach_clk_sig
+
+
+class Platform(LatticePlatform):
+    default_clk_name = "osch_clk"
+    default_clk_period = 1000.0/15.65
+    osch_clk_period = 1000.0/15.65
+    # osch_clk_period is the variable used to set the OSCH period.
```

Comment:
If it's a variable, then maybe it should be an argument to ``__init__`` or some other method, not a class attribute.

Reply:
What if we assume `osch_clk_period` is immutable once set, rather than "set at will"? Out of curiosity, why are `default_clk_name` and `default_clk_period` class vars? In a similar vein, once they are set there should be no reason to modify them?

Reply:
> why are default_clk_name and default_clk_period class vars (in all platform files)?

They are just the clock that Migen uses for the ``sys`` domain on this platform when the user doesn't specify a ``sys`` domain - and nothing else.

Can you change your patch based on this consideration?